cirq/circuits/circuit_gate.py

```diff
@@ -0,0 +1,384 @@
+        if protocols.is_measurement(new_gate) and self._repetitions != 1:
+            raise NotImplementedError(
+                'Measurements cannot be added to looped circuits.')
+        new_op = self.copy()
+        new_op._gate = new_gate
+        return new_op
+
+    def with_measurement_key_mapping(self, key_map: Dict[str, str]
+                                    ) -> 'CircuitOperation':
+        """Returns a copy of this operation with an updated key mapping.
+
+        Parameter values and repetitions are preserved. The provided key_map
+        is composed with the existing map, so calling this function with
+        key_map = {'b': 'c'} on a CircuitOperation with map {'a': 'b'} will
+        result in a final map of {'a': 'c'}.
+        """
+        new_op = self.copy()
+        new_op._measurement_key_map = {
+            k: (key_map[v] if v in key_map else v)
+            for k, v in self.measurement_key_map.items()
+        }
+        new_op._measurement_key_map.update({
+            key: val
+            for key, val in key_map.items()
+            if key not in self.measurement_key_map.values()
+        })
+        return new_op
+
+    def _with_measurement_key_mapping_(self, key_map: Dict[str, str]
+                                      ) -> 'CircuitOperation':
+        return self.with_measurement_key_mapping(key_map)
+
+    def with_params(self, param_values: Dict[sympy.Symbol, Any]
+                   ) -> 'CircuitOperation':
+        """Returns a copy of this operation with an updated key mapping.
+
+        Key mappings and repetitions are preserved. The provided param_values
+        are composed with the existing values, so calling this function with
+        param_values = {'b': 1} on a CircuitOperation with map {'a': 'b'} will
+        result in a final map of {'a': 1}.
+        """
+        new_op = self.copy()
+        new_op._param_values = {
+            key: (val if not protocols.is_parameterized(val) else
+                  protocols.resolve_parameters(val, param_values))
+            for key, val in self.param_values.items()
+        }
+        new_op._param_values.update({
+            key: val
+            for key, val in param_values.items()
+            if key not in self.param_values.values()
+        })
+        return new_op
+
+    def repeat(self, repetitions: int) -> 'CircuitOperation':
+        """Returns a copy of this operation repeated 'repetitions' times.
+
+        Key mappings and parameter values are preserved.
+        """
+        if not isinstance(repetitions, int):
+            raise TypeError('Only integer repetitions are allowed.')
+        if protocols.is_measurement(self.circuit):
+            raise NotImplementedError(
+                'Loops over measurements are not supported.')
+        new_op = self.copy()
+        new_op._repetitions *= repetitions
+        if self.gate.exp_modulus is not None:
+            # Map repetitions to [-exp_modulus / 2, exp_modulus / 2)
```

Comment:
If we do include something like `exp_modulus`, I'd suggest that there should be a way to specify whether or not it should be inverted. Repetitions between `[0, exp_modulus)` might make more sense in some cases. Also, for the half interval it might make more sense to be inclusive on the upper instead of lower bound, so that circuits with modulus 2 never need to be inverted, for example.

Reply:
My current plan is to package this behavior (exp_modulus, modulus validation, and invertibility) into the `repeat` method instead. That should ensure users (or optimizers) are explicit about using those behaviors.